### PR TITLE
Tinsel: Basic 3D rendering of the main character for Discworld Noir.

### PR DESCRIPTION
**Pull request by peterkohaut**
These patches add basic 3D rendering of the main character model, the only model in Discworld Noir.
The character model loads and renders successfully in the first scene, but there is no movement or animation, which requires additional reverse-engineering of the ground plane and the movers.

For 3D rendering, I used TinyGL because the Tinsel engine can draw the 3D model in the middle of the drawing of the other game objects, which might even be transparent. There is no easy way to use OpenGL directly due to this. 

These changes are limited to Discworld Noir. I tested the changes with Discworld 1 & 2, and they work as before.

I had these patches in my fork for some time and cleaned them up. It would be nice to get some eyes on them.

### CODE REVIEW

**Comment by sev-** [comment]
oh, wow!

File: engines/tinsel/background.cpp [hunk 1]
```diff
@@ -218,22 +219,52 @@ void Background::DrawBackgnd() {
 
 	// redraw all playfields within the clipping rectangles
 	const RectList &clipRects = GetClipRects();
-	for (RectList::const_iterator r = clipRects.begin(); r != clipRects.end(); ++r) {
-		// clear the clip rectangle on the virtual screen
-		// for each background playfield
-		for (unsigned int i = 0; i < _pCurBgnd->fieldArray.size(); i++) {
-			Common::Rect rcPlayClip;	// clip rect for this playfield
 
+	// Noir 3D model's cliprect is updated the rendering.
```
Comment (bluegr) [inline engines/tinsel/background.cpp:223]:
```suggestion
	// Noir 3D model's cliprect is updated for rendering.
```

File: engines/tinsel/background.cpp [hunk 2]
```diff
@@ -218,22 +219,52 @@ void Background::DrawBackgnd() {
...
 
-			if (IntersectRectangle(rcPlayClip, pPlay->rcClip, *r))
-				// redraw all objects within this clipping rect
-				UpdateClipRect(&pPlay->pDispList, &ptWin,	&rcPlayClip);
+			for (OBJECT* pObj = pPlay->pDispList; pObj != NULL; pObj = pObj->pNext) {
+				if (pObj->flags & DMA_3D) {
+					_vm->_spriter->Draw(0, 0, 0, 0, 0);
+					//AddClipRect(SpriterGetModelRect());
+					//AddClipRect(SpriterGetShadowRect());
+					//MergeClipRect();
```
Comment (bluegr) [inline engines/tinsel/background.cpp:238]:
```suggestion
```

File: engines/tinsel/cliprect.cpp [hunk 3]
```diff
@@ -197,116 +197,127 @@ void MergeClipRect() {
...
-			currentObj.isRLE      = pObj->isRLE;
-			currentObj.colorFlags = pObj->colorFlags;
-		}
-		currentObj.constant       = pObj->constant;
-		currentObj.hBits          = pObj->hBits;
+		if (pObj->height <= vclip)
+			// object totally clipped vertically - ignore
+			return;
+
+		// set clip bit in objects flags
```
Comment (bluegr) [inline engines/tinsel/cliprect.cpp:286]:
```suggestion
		// set clip bit in object flags
```

File: engines/tinsel/cliprect.cpp [hunk 4]
```diff
@@ -197,116 +197,127 @@ void MergeClipRect() {
...
+		currentObj.flags = pObj->flags | DMA_CLIP;
+	} else {	// object is not clipped - copy flags
+		currentObj.flags = pObj->flags;
+	}
+
+	// copy objects properties to local object
```
Comment (bluegr) [inline engines/tinsel/cliprect.cpp:292]:
```suggestion
	// copy object properties to local object
```

File: engines/tinsel/noir/spriter.cpp [hunk 5]
```diff
@@ -0,0 +1,1215 @@
...
+#endif
+	_textureGenerated = true;
+}
+
+void Spriter::SetPalette(SCNHANDLE hPalette) {
+	const
+	uint32 paletteHeaderSize = 2 + 2;
+	uint32 paletteBodySize = PALETTE_COUNT * 256 * sizeof(uint16);
+
+	/* Select only one palette as they are sorded by light intensity */
```
Comment (bluegr) [inline engines/tinsel/noir/spriter.cpp:442]:
```suggestion
	/* Select only one palette as they are sorted by light intensity */
```

File: engines/tinsel/noir/spriter.cpp [hunk 6]
```diff
@@ -0,0 +1,1215 @@
...
+	// 	_modelMain.position.z() + sheerz * 10.0f
+	// );
+
+	// _modelShadow.rotation.set(
+	// 	_modelMain.rotation.x(),
+	// 	_modelMain.rotation.y() + 180.0f,
+	// 	_modelMain.rotation.z()
+	// );
+
+	// RenderModel(_modelShadow);
```
Comment (bluegr) [inline engines/tinsel/noir/spriter.cpp:582]: Why is this code commented out? Either put a TODO here, or remove it altogether

File: engines/tinsel/noir/spriter.cpp [hunk 7]
```diff
@@ -0,0 +1,1215 @@
...
+}
+
+void Spriter::RenderMesh(Mesh& mesh, Vectors& vertices, Vectors &normals) {
+	for(auto& part : mesh.parts) {
+		switch(part.type) {
+		case MESH_PART_TYPE_COLOR:
+			RenderMeshPartColor(part, vertices, normals);
+			break;
+		case MESH_PART_TYPE_SOLID:
+			// This is just use white color
```
Comment (bluegr) [inline engines/tinsel/noir/spriter.cpp:984]:
```suggestion
			// This just uses white color
```

File: engines/tinsel/noir/spriter.cpp [hunk 8]
```diff
@@ -0,0 +1,1215 @@
...
+	LoadGBL(modelName);
+	LoadRBH(modelName, _modelMain.hunks);
+	LoadVMC(modelName);
+
+	InitModel(_modelMain, _meshMain, _animMain, MODEL_HAS_TRANSLATION_TABLE | MODEL_HAS_ROTATION_TABLE);
+
+	// for (uint i = 0; i < _animMain.size(); ++i) {
+	// 		update max frame
+	// }
+
```
Comment (bluegr) [inline engines/tinsel/noir/spriter.cpp:1073]: Please add a TODO here

File: engines/tinsel/noir/spriter.cpp [hunk 9]
```diff
@@ -0,0 +1,1215 @@
...
+	tglFrontFace(TGL_CW);
+
+	tglMatrixMode(TGL_MODELVIEW);
+	tglLoadIdentity();
+
+	// tglRotatef(_view.rotation.x(), 1.0f, 0.0f, 0.0f);
+	// tglRotatef(_view.rotation.y(), 0.0f, 1.0f, 0.0f);
+	// tglRotatef(_view.rotation.z(), 0.0f, 0.0f, 1.0f);
+	// tglTranslatef(-_view.position.x(), -_view.position.y(), -_view.position.z());
+
```
Comment (bluegr) [inline engines/tinsel/noir/spriter.cpp:1144]: Please add a TODO here

**Review by bluegr — APPROVED** [review]
Great work overall - well done! :)

There are some places where commented out code has been added, without any TODO or other explanation. Please, update these accordingly.

File: engines/tinsel/tinsel.cpp [hunk 10]
```diff
@@ -1021,7 +1027,16 @@ Common::Error TinselEngine::run() {
...
-		_screenSurface.create(width, 432, noirFormat);
+#if defined(USE_TINYGL)
+		TinyGL::createContext(width, 432, noirFormat, 256, false, false);
+		TinyGL::getSurfaceRef(_screenSurface);
+
+		_spriter = new Spriter();
+		_spriter->Init(width, 432);
+#else
+		error("Discworld Noir needs ScummVM with TinyGL enabled");
+		return Common::kUnknownError;
```
Comment (ccawley2011) [inline engines/tinsel/tinsel.cpp:1038]:
```suggestion
		return Common::Error(Common::kUnsupportedGameidError, _s("Discworld Noir needs ScummVM with TinyGL enabled"));
```

File: engines/tinsel/tinsel.cpp [hunk 11]
```diff
@@ -1021,7 +1027,16 @@ Common::Error TinselEngine::run() {
 
 		initGraphics(width, height, &noirFormat);
 
-		_screenSurface.create(width, 432, noirFormat);
+#if defined(USE_TINYGL)
+		TinyGL::createContext(width, 432, noirFormat, 256, false, false);
```
Comment (ccawley2011) [inline engines/tinsel/tinsel.cpp:1031]: Would it be possible to specify a lower value for the `drawCallMemorySize` parameter than the default one?

**Pull request merged**